nisapi/clean.py

```diff
@@ -0,0 +1,253 @@
+
+        # Verify that indicator type "up-to-date" has only one value ("yes")
+        assert clean.filter(pl.col("indicator_type") == pl.lit("up-to-date")).pipe(
+            col_values_in, "indicator_value", ["yes"]
+        )
+
+        # check that "Yes" and "Received a vaccination" are the same thing, so that
+        # we can drop "Up to Date"
+        assert (
+            clean.filter(
+                pl.col("indicator_value").is_in(["yes", "received a vaccination"])
+            )
+            .drop("indicator_type")
+            .pivot(on="indicator_value", values=["estimate", "ci_half_width_95pct"])
+            .select(
+                (
+                    (
+                        pl.col("estimate_yes")
+                        == pl.col("estimate_received a vaccination")
+                    )
+                    & (
+                        pl.col("ci_half_width_95pct_yes")
+                        == pl.col("ci_half_width_95pct_received a vaccination")
+                    )
+                ).all()
+            )
+            .item()
+        )
+
+        clean = clean.filter(
+            pl.col("indicator_type") == pl.lit("4-level vaccination and intent")
+        )
+
+        return clean
+
+    validate(clean)
+    return clean
+
+
+def rename_indicator_columns(df: pl.DataFrame) -> pl.DataFrame:
+    """
+    Make "indicator" follow the same logic as "geography" and
+    "demographic", with "type" and "value" columns
+    """
+    return df.rename(
+        {
+            "geographic_level": "geographic_type",
+            "geographic_name": "geographic_value",
+            "demographic_level": "demographic_type",
+            "demographic_name": "demographic_value",
+            "indicator_label": "indicator_type",
+            "indicator_category_label": "indicator_value",
+        }
+    )
+
+
+def col_values_in(df: pl.DataFrame, col: str, values: str) -> bool:
+    """All values of `df` in column `col` are in `values`?"""
+    return df[col].is_in(values).all()
+
+
+def remove_near_duplicates(df: pl.DataFrame, tolerance: float = 1e-3) -> pl.DataFrame:
+    value_cols = ["estimate", "ci_half_width_95pct"]
+    group_cols = set(df.columns) - set(value_cols)
+
+    nearly_dup_rows = df.drop(value_cols).is_duplicated()
+
+    # assert that there are only 2 of each of these
+    assert (
+        df.filter(nearly_dup_rows)
+        .select(group_cols)
+        .group_by(pl.all())
+        .count()["count"]
+        == 2
+    ).all()
+
+    # assert that the estimates and CIs in these groups are similar to
+    # one another, within some small margin
+    assert (
+        df.filter(nearly_dup_rows)
+        .group_by(group_cols)
+        .agg((pl.col(value_cols).pipe(lambda x: x.max() - x.min())))
+        .select((pl.col(value_cols) < tolerance).all())
+        .select(pl.all_horizontal(value_cols))
+        .item()
+    )
+
+    # drop the nearly-duplicate rows
+    return df.filter(nearly_dup_rows.not_())
```

Comment:
Looks to me like this deletes every copy of nearly-duplicate rows. Shouldn't we keep exactly one copy of nearly-duplicate rows?

Reply:
Good catch; see #22